Modernize and Refactor Tour components

ahmedhamidawan opened the pull request:
This also aims to fix minor bugs, and streamline the tours interface (properly terminating in case of an error during a tour for e.g.).

| 🔴 Before: Improper error handling in tours |
| ---- |
| <video src="https://github.com/user-attachments/assets/a5059bb9-66ed-4bc5-bf9d-737356f00411" /> |
| Example here is: Tour expects a tool panel but it was started with the panel collapsed. The error is caught but not propagated up properly (we weren't using the `reject` in `runTour`). |

| 🟢 Fixed: Error displayed in tour modal |
| ---- |
| <video src="https://github.com/user-attachments/assets/0ff4795a-97d3-4dc4-a681-47906bb19677" /> |
| Any errors occurring during a tour are properly displayed, and user can even go back and try the same step again. _(In this case, it's because of the tool panel not expanded, once we do that, we can continue the tour unlike before.)_  |

### Improved routing
By mounting `TourRunner` in `App.vue`, we can have tours that navigate across all app routes easily. Note that page reloads still clear the store.

https://github.com/user-attachments/assets/cc3eca59-491c-4eaf-8c1a-d9b99104332d

<details>

  <summary>
    <h3>(Outdated)
      An approach I tried to ensure these errors never happen in the first place, but bailed on it as mentioned in comments below
    </h3>
  </summary>

| A way to fix the no-selector problem here ^ |
| ---- |
| <video src="https://github.com/user-attachments/assets/c090f678-1f7d-4577-a99a-f1050fce1d25" /> |
| I've tried this approach of adding a `prerequisite` key to the `TourStep`s, which is a CSS selector we try to click if the expected component/element doesn't already exist. _(In the example here, if tool panel doesn't exist, we click the Tools activity which is the `prerequisite` here.)_ |

**Note: Ignore the error at the end here ^. It's from testing and I had set the `const attempts` in `runTour` to 1.**

</details>

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
Refactor Tour components to composition API and TS

This also aims to fix minor bugs, and streamline the tours interface (properly terminating in case of an error during a tour for e.g.).

ahmedhamidawan committed:
use a ref to control modal visibility

It seems to work better with reactivity

ahmedhamidawan committed:
add modal ok function

ahmedhamidawan committed:
add `routePush` function since `vue-router` is not available in `Tour.vue`

The component is mounted in `runTour` and somehow doesn't have access to the router when mounted.

ahmedhamidawan committed:
catch and display errors in Tour

jmchilton commented on config/plugins/tours/core.galaxy_ui.yaml:
```diff
@@ -51,20 +51,25 @@ steps:
       component: tools.search
       intro: "You can search for tools by keywords."
       textinsert: 'filter'
+      prerequisite: "#activity-tools"
```
Is there anyway I can convince you these should be components instead of raw CSS selectors (https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tours/_impl.py#L47) or to implement these prerequisites in the testing framework (https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/selenium/navigates_galaxy.py#L2478).

ahmedhamidawan replied:
Absolutely, just wanted to tease the idea of these "prerequisites" first. 

We should definitely use components instead

ahmedhamidawan replied:
Not doing prerequisites anymore as mentioned below

guerler commented on client/src/components/Tour/runTour.js:
```diff
@@ -27,22 +27,34 @@ function getElement(selector) {
+                resolve();
+            } else if (tries > 0) {
+                // First, try to click the prerequisite element if it exists
+                if (prerequisite) {
```
Does this not cause multiple clicks on the `prerequisite` element i.e. keeps opening and closing the tool panel, maybe a conditional step would work better. Thank you for working on this.

ahmedhamidawan replied:
Ah thats a good point 👉 
Will fix that

ahmedhamidawan replied:
Actually, the whole point of adding this is to try to make the expected element appear (e.g. open the tool panel) _if it's not already visible_ (opened in the case of the panel).

So, since in the lines above this, we already confirm that the expected element (e.g.: panel) isn't open:
- If it is already visible, we `resolve()`
- If not visible yet, try to make it appear then by clicking the `prerequisite` (in the case of the tool panel, the prerequisite is the tools activity)

What I have here should be fine?

ahmedhamidawan replied:
Not doing prerequisites anymore as mentioned below

ahmedhamidawan committed:
center tour popper for an element that is now not visible

This is the case where there was a popper hovering for an element, which is then not there anymore. Without this change, the popper appears at coordinates (20,20) on top left; which is worse than just centering the popper for an invisible element.

ahmedhamidawan committed:
make `runTour` legacy; use `Tour` in `TourList` in `Analysis`

Handles mounting and unmounting tours really well in `Analysis`.

ahmedhamidawan committed:
rename tour end button from "Cancel" to "End Tour"

ahmedhamidawan committed:
add waiting for element indicator when waiting on a CSS element

ahmedhamidawan committed:
modernize `TourStep` component

guerler reviewed (approved):
This looks good to me. I am just wondering if we should move the route watching in the Analysis module back into the router, where it might belong. Overall this is great, thank you.

ahmedhamidawan committed:
use GLink and API types in `TourList`

ahmedhamidawan committed:
add `tourStore` that keeps track of a running tour

We mount `TourRunner` in `App.vue` instead of `Analysis.vue`, because this way we can have a tour run across all routes (even workflow editor, published views etc.).

Note that an entire page reload resets the store still...

Co-authored-by: Aysam Guerler <aysam.guerler@gmail.com>

guerler commented on client/src/entry/analysis/App.vue:
```diff
@@ -124,7 +131,16 @@ export default {
+
+                // if we are on a tour route, start a tour if it wasn't already started or change tours
+                if ("tourId" in route.params) {
+                    const routeTourId = route.params.tourId === "null" ? null : route.params.tourId;
```
Under which circumstances does the tourId become a string with value "null"?

ahmedhamidawan replied:
Oh good catch, this was from when I had tourStore.currentTourId as a string in `useUserLocalStorage` (some reactivity issue there would set the value to "null".

We can just remove this.

ahmedhamidawan committed:
prevent localStorage `persistentRef` from being string null/undefined

ahmedhamidawan committed:
cache tours to make `runTour` also work

This exists solely for the `tour_generator` webhook as of now. We are going to be refactoring and changing that from a webhook to core functionality, so we have also commented out the selenium that tests "Tool Describing Tours".

guerler reviewed (approved):
Thank you for reworking the tour state handling. Using session storage will allow us to persist tours across page reloads. I also ok with temporarily disabling auto-generated tool tour tests in this PR, assuming we follow up soon and move that webhook into core, since it was brittle and prone to breaking.
